content/blog/running-self-hosted-agents-in-azure-devops-using-an-azure-ad-access-token.md

```diff
@@ -0,0 +1,123 @@
+
+It doesn't really matter why, you (like me) probably had second thoughts when you saw that the recommended way of getting a Self Hosted Agent to connect with your Azure Devops organization is by using a Personal Access Tokens (PAT).
+
+PAT's might be very easy to create, but in my experience, they are a pain to manage. Rotating a PAT doesn't take long (but cannot trivialy be automatically rotated in a non-interactive setting), but they are bound to a single person, making them unreliable to use in any business-critical applications. Also, if you scope them correctly you end up having a lot of them, making a small amount of manual work become an error-prone and tedious operation, making it very attractive to just have a single PAT that is way overprovisioned for what it was actually meant to do.
+
+Luckily, there's another [way](https://devblogs.microsoft.com/devops/introducing-service-principal-and-managed-identity-support-on-azure-devops/). Since late March, Microsoft has made it possible for organizations that use Azure AD (soon to be renamed to Microsoft Entra ID, for reasons still unclear to me and everyone else) to use App Registrations and their resulting Service Principals to create Azure AD Access Tokens that can be used in place of PAT's.
+
+Sounds great! How though? Well that's what we're here for, so let's get going.
+
+What you're going to need is:
+- An Azure subscription with Azure AD enabled
+- An Azure Devops (ADO) organization that is connected to a Directory inside aforementioned Azure Subscription
+- A project inside said ADO organization in which you're permitted to create new Self Hosted Agent Pools and set Organization Level User Permissions.
+- Something to run a Self Hosted agent on. For more inspiration, check out [this article](https://learn.microsoft.com/en-us/azure/devops/pipelines/agents/agents?view=azure-devops&tabs=browser#install)
+  - Pro tip: If you're planning on running a Self Hosted Agent on a Debian VM, I've found it way easier to just run the agent as a docker/podman container instead of getting the Linux agent software working. There seems to be some unsolved bug in there that causes the agent not to be able to connect to Azure Devops, which is quite a limitation(!).
+
+First up, as mentioned before, we're going to need an Azure environment that has Azure AD enabled. Log into your portal and head to "App Registrations". Register a new app, and give it a fitting name. I named mine "self-hosted-agent-pool-sp".
+
+{{<img src="/img/blog/create-app-registration.png" class="img-fluid" title="Create app registration" >}}
+
+Next up, go to API Permissions and add the Azure Devops user_impersonation Delegated permission. This will allow the Service Principal to call the Azure Devops REST API, which is what the Agent software is doing when registering itself as a new agent inside a given agent pool.
```

Comment:
Should `user_impersonation` be in code syntax for readability?

Reply:
Yeah, it should. Will fix!